File: server/src/services/partner.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { Partner } from 'src/database.js';
import { AuthDto } from 'src/dtos/auth.dto.js';
import { PartnerCreateDto, PartnerResponseDto, PartnerSearchDto, PartnerUpdateDto } from 'src/dtos/partner.dto.js';
import { mapUser } from 'src/dtos/user.dto.js';
import { Permission, SharingDirection } from 'src/enum.js';
import { PartnerIds } from 'src/repositories/partner.repository.js';
import { BaseService } from 'src/services/base.service.js';

@Injectable()
export class PartnerService extends BaseService {
  async create(auth: AuthDto, { sharedWithId }: PartnerCreateDto): Promise<PartnerResponseDto> {
    const partnerId: PartnerIds = { sharedById: auth.user.id, sharedWithId };
    const exists = await this.partnerRepository.get(partnerId);
    if (exists) {
      throw new BadRequestException(`Partner already exists`);
    }

    const user = await this.userRepository.get(sharedWithId, {});
    if (!user) {
      this.logger.debug('Partner creation failed: user not found');
      throw new BadRequestException('Invalid user');
    }

    const partner = await this.partnerRepository.create(partnerId);
    return this.mapPartner(partner, SharingDirection.SharedBy);
  }

  async remove(auth: AuthDto, sharedWithId: string): Promise<void> {
    const partnerId: PartnerIds = { sharedById: auth.user.id, sharedWithId };
    const partner = await this.partnerRepository.get(partnerId);
    if (!partner) {
      throw new BadRequestException('Partner not found');
    }

    await this.partnerRepository.remove(partnerId);
  }

  async search(auth: AuthDto, { direction }: PartnerSearchDto): Promise<PartnerResponseDto[]> {
    const partners = await this.partnerRepository.getAll(auth.user.id);
    const key = direction === SharingDirection.SharedBy ? 'sharedById' : 'sharedWithId';
    return partners
      .filter((partner): partner is Partner => !!(partner.sharedBy && partner.sharedWith)) // Filter out soft deleted users
      .filter((partner) => partner[key] === auth.user.id)
      .map((partner) => this.mapPartner(partner, direction));
  }

  async update(auth: AuthDto, sharedById: string, dto: PartnerUpdateDto): Promise<PartnerResponseDto> {
    await this.requireAccess({ auth, permission: Permission.PartnerUpdate, ids: [sharedById] });
    const partnerId: PartnerIds = { sharedById, sharedWithId: auth.user.id };

    const entity = await this.partnerRepository.update(partnerId, { inTimeline: dto.inTimeline });
    return this.mapPartner(entity, SharingDirection.SharedWith);
  }

  private mapPartner(partner: Partner, direction: SharingDirection): PartnerResponseDto {
    // this is opposite to return the non-me user of the "partner"
    const sharedUser = direction === SharingDirection.SharedBy ? partner.sharedWith : partner.sharedBy;
    const user = mapUser(sharedUser);

    return { ...user, inTimeline: partner.inTimeline };
  }
}

File: mobile/lib/domain/services/partner.service.dart
import 'package:immich_mobile/constants/enums.dart';
import 'package:immich_mobile/domain/models/user.model.dart';
import 'package:immich_mobile/infrastructure/repositories/partner.repository.dart';
import 'package:immich_mobile/infrastructure/repositories/user.repository.dart';
import 'package:immich_mobile/repositories/partner_api.repository.dart';
import 'package:stream_transform/stream_transform.dart';

class PartnerService {
  final UserRepository _userRepository;
  final PartnerRepository _partnerRepository;
  final PartnerApiRepository _partnerApiRepository;

  const PartnerService(this._userRepository, this._partnerRepository, this._partnerApiRepository);

  Stream<Iterable<User>> getCandidates(String userId) {
    final userStream = _userRepository.getAll();
    final partnerStream = _partnerRepository.search(userId, .sharedBy);

    return userStream.combineLatest(partnerStream, (users, partners) {
      final partnersSet = partners.map((partner) => partner.id).toSet();
      return users.where((user) => user.id != userId && !partnersSet.contains(user.id));
    });
  }

  Stream<Iterable<Partner>> search(String userId, PartnerDirection direction) =>
      _partnerRepository.search(userId, direction);

  Future<void> update({required String sharedById, required String sharedWithId, required bool inTimeline}) async {
    await _partnerApiRepository.update(sharedById, inTimeline: inTimeline);
    await _partnerRepository.update(sharedById: sharedById, sharedWithId: sharedWithId, inTimeline: inTimeline);
  }

  Future<void> create({required String sharedById, required String sharedWithId, bool inTimeline = false}) async {
    await _partnerApiRepository.create(sharedWithId);
    await _partnerRepository.create(sharedById: sharedById, sharedWithId: sharedWithId, inTimeline: inTimeline);
  }

  Future<void> delete({required String sharedById, required String sharedWithId}) async {
    await _partnerApiRepository.delete(sharedWithId);
    await _partnerRepository.delete(sharedById: sharedById, sharedWithId: sharedWithId);
  }
}
